File: app/src/main/java/com/letianpai/robot/control/callback/TemperatureUpdateCallback.java
package com.letianpai.robot.control.callback;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by liujunbin
 */

public class TemperatureUpdateCallback {
    public static final int HIGH_TEMP = 90;
    public static final int TARGET_TEMP = 75;
    private boolean isHighTemp = false;

    private float temp;
    private List<TemperatureUpdateListener> mTemperatureUpdateListener = new ArrayList();

    private static class TemperatureUpdateCallBackHolder {
        private static TemperatureUpdateCallback instance = new TemperatureUpdateCallback();
    }

    private TemperatureUpdateCallback() {

    }

    public static TemperatureUpdateCallback getInstance() {
        return TemperatureUpdateCallBackHolder.instance;
    }

    public interface TemperatureUpdateListener {
        void onTemperatureUpdate(float temp);
    }

    public void registerTemperatureUpdateListener(TemperatureUpdateListener listener) {
        if (mTemperatureUpdateListener != null) {
            mTemperatureUpdateListener.add(listener);
        }
    }

    public void unregisterTemperatureUpdateListener(TemperatureUpdateListener listener) {
        if (mTemperatureUpdateListener != null) {
            mTemperatureUpdateListener.remove(listener);
        }
    }

    public void setTemperature(float temp) {
        this.temp = temp;
        if (temp >= HIGH_TEMP){
            isHighTemp = true;

        }else if (temp <= TARGET_TEMP){
            isHighTemp = false;

        }else if (temp > TARGET_TEMP && temp < HIGH_TEMP){

        }else{
            isHighTemp = false;
        }

        for (int i = 0; i < mTemperatureUpdateListener.size(); i++) {
            if (mTemperatureUpdateListener.get(i) != null) {
                mTemperatureUpdateListener.get(i).onTemperatureUpdate(temp);
            }
        }
    }

    public boolean isInHighTemperature(){
        return isHighTemp;
    }

    public float getTemp() {
        return temp;
    }
}

File: app/src/main/java/com/letianpai/robot/control/callback/TemperatureUpdateCallback.kt
package com.letianpai.robot.control.callback

/**
 * Created by liujunbin
 */
class TemperatureUpdateCallback private constructor() {
    var isInHighTemperature: Boolean = false
        private set

    var temp: Float = 0f
        private set
    private val mTemperatureUpdateListener: MutableList<TemperatureUpdateListener?> =
        ArrayList()

    private object TemperatureUpdateCallBackHolder {
        val instance: TemperatureUpdateCallback = TemperatureUpdateCallback()
    }

    interface TemperatureUpdateListener {
        fun onTemperatureUpdate(temp: Float)
    }

    fun registerTemperatureUpdateListener(listener: TemperatureUpdateListener?) {
        if (mTemperatureUpdateListener != null) {
            mTemperatureUpdateListener.add(listener)
        }
    }

    fun unregisterTemperatureUpdateListener(listener: TemperatureUpdateListener?) {
        if (mTemperatureUpdateListener != null) {
            mTemperatureUpdateListener.remove(listener)
        }
    }

    fun setTemperature(temp: Float) {
        this.temp = temp
        if (temp >= HIGH_TEMP) {
            isInHighTemperature = true
        } else if (temp <= TARGET_TEMP) {
            isInHighTemperature = false
        } else if (temp > TARGET_TEMP && temp < HIGH_TEMP) {
        } else {
            isInHighTemperature = false
        }

        for (i in mTemperatureUpdateListener.indices) {
            if (mTemperatureUpdateListener[i] != null) {
                mTemperatureUpdateListener[i]!!.onTemperatureUpdate(temp)
            }
        }
    }

    companion object {
        const val HIGH_TEMP: Int = 90
        const val TARGET_TEMP: Int = 75
        val instance: TemperatureUpdateCallback
            get() = TemperatureUpdateCallBackHolder.instance
    }
}
